proposals/scripting-support.md

```diff
@@ -0,0 +1,434 @@
+    
+    val val1: V1 by initOnce // for all vals/vars defined in the script body
+    
+    fun fn1(...): R1 {} // for all funcs defined in the script body
+    
+    class Cl1(...) {} // for all classes/objects defined in the script body
+    
+    var param1: P1 = additionalArgs["param1"] // for all additionalParams, the types are defined by `compilationConfig` extracted by Preprocessor 
+    
+    [suspend] fun <SAM name>(lambdaParams...): RetVal {
+        with(implicitReceivers) {
+            ...
+        }
+    }
+}
+```
+
+In the REPL mode `implicitReceivers` will contain all previous script lines (the order is significant).
+
+The `implicitReceivers` may also contain compiled external scripts objects.
+
+*(The `with (implicitReceivers)` wrapping is needed in all methods and initializers defined in the class.)*
+
+The additional parameters may include params extracted by preprocessor (`@file:param("name", String::class)`) or could 
+be used for bindings mapping in JSR-223-like implementations ([KT-18917](https://youtrack.jetbrains.com/issue/KT-18917)).  
+
+### Script definition
+
+**Script Definition** is a configuration entity that combines platform, preprocessor, and selector in one entity 
+for simplified discovery and configuration. 
+
+The definition is a kotlin class declaration annotated with specific annotations. The class should be defined in the SAM 
+notation, and the single abstract method will be used by the platform to compile the script body into.  
+*(avoiding usage of the word "Template", since it seems confuses people)* :
+
+```
+@ScriptDefinition(
+    platform = JVMScriptPlatform::class,
+    preprocessor = MyScriptPreprocessor::class,
+    selector = MyScriptSelector::class,
+)
+abstract class MyScript(project: Project, val name: String) {
+    fun helper1() { ... } 
+    
+    [suspend] abstract fun samFn(params...): R // SAM notation
+}
+```
+
+Alternatively, individual annotations for some elements are possible for specific cases:
+
+```
+@ScriptByFileNameSelector("*.gradle.kts")
+```
+
+### Compilation Platform
+
+```
+interface ScriptingPlatform<CompilationConfig> {
+    fun compile(scriptSource, scriptBaseClass: KClass<Scritpt>, compilationConfig: CompilationConfig): KClass<out Script>
+    fun <getIdeData>(scriptSource, scriptBaseClass: KClass<Script>, compilationConfig: CompilationConfig): <IdeData>
+}
+```
+
+### Executor
+
+```
+interface ScriptExecutor<Script, CompilationConfig> {
+    fun createInstance(compiledScript: KClass<out Script>, compilationConfig: CompilationConfig, params...): Script
+    fun run(scriptObject: Script, compilationConfig: CompilationConfig): Unit
+}
+```
+
+### Preprocessor
+
+```
+interface ScriptPreprocessor<CompilationConfig> {
+    fun preprocess(scriptSource): CompilationConfig 
+}
+```
+
+### Selector
+
+```
+interface ScriptSelector {
+    fun isScript(scriptSource): boolean 
```

Comment:
I agree. I was playing with `isKnownScript` name as well, but it looked cumbersome. May be it is the better one anyway.